Add support for floating point operations

smortex opened the pull request:
This PR is about supporting math expression using floating-point values.  The code is published for review for now, because some decisions might need to be discussed.  I plan to amend the commits to address the issues and comments raised by the syslog-ng community, and remove the `[Do not merge]` tag when I consider the changes to be ready.

Fixes #2734

kira-syslogng commented:
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

kira-syslogng commented:
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

smortex commented on modules/basicfuncs/numeric-funcs.c:
```diff
@@ -21,11 +21,66 @@
+    gint64 gint64;
+    gdouble gdouble;
+  } value_data;
+};
```
`value` is an awful name.  Any suggestion for a better name is welcome :wink:, the idea here is to be able to distinguish if we parsed and integer or a float, and store the value in an easy to manipulate format.

Kokan commented:
@kira-syslogng ok to test

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

Kokan commented:
@smortex I brough up this feature to the team, and we want to discuss it. It may take some time, but I hope not later than next Wednesday.

smortex commented:
Sure, thanks you for taking care of it!  I started "the easy way" and I guess a lot of topics will be raised, such as:

* should `sum`, `minimum`, `maximum` & co be updated to support floats;
* should the system be reworked to minimize the impact of roundings when the result of an operation is used in another operation;
* etc.

Maintainers of the project can push commits to my branch, feel free to do so, but please send me a heads-up in this case so that I do not `git push -f` and loose your changes unexpectedly :wink:

Kokan commented on modules/basicfuncs/numeric-funcs.c:
```diff
@@ -21,11 +21,66 @@
+    return FALSE;
+
+  return TRUE;
+}
```
[optional]
```c
gboolean
parse_value(const char *str, struct value *value)
{
  if (parse_dec_number(str, &(value->value_data.gint64))) 
   {
       value->value_type = VALUE_TYPE_GINT64;
       return TRUE;
    }

 if (parse_float(str, &(value->value_data.gdouble)))
   {
       value->value_type = VALUE_TYPE_GDOUBLE;
       return TRUE;
    }

  return FALSE;
}
```

Kokan commented on modules/basicfuncs/numeric-funcs.c:
```diff
@@ -56,83 +111,204 @@ tf_num_parse(gint argc, GString *argv[],
+    {
+      gdouble res = value_as_double(n) + value_as_double(m);
+      g_string_append_printf(result, "%f", res);
+    }
```
Please extract this part into a function (as used in many places) like `format_number`.

Kokan commented on modules/basicfuncs/numeric-funcs.c:
```diff
@@ -56,83 +111,204 @@ tf_num_parse(gint argc, GString *argv[],
+      return;
+    }
+
+  g_string_append_printf(result, "%ld", (gint64) round(value_as_double(n)));
```
Please use `format_int64_padded` instead of printf.

Kokan commented on modules/basicfuncs/numeric-funcs.c:
```diff
@@ -56,83 +111,204 @@ tf_num_parse(gint argc, GString *argv[],
+      return;
+    }
+
+  g_string_append_printf(result, "%ld", (gint64) ceil(value_as_double(n)));
```
Please use `format_int64_padded` instead of printf.

Kokan commented on modules/basicfuncs/numeric-funcs.c:
```diff
@@ -56,83 +111,204 @@ tf_num_parse(gint argc, GString *argv[],
+      return;
+    }
+
+  g_string_append_printf(result, "%ld", (gint64) floor(value_as_double(n)));
```
Please use `format_int64_padded` instead of printf.

Kokan commented on modules/basicfuncs/tests/test_basicfuncs.c:
```diff
@@ -219,6 +219,19 @@ Test(basicfuncs, test_numeric_funcs)
+  assert_template_format("$(round 2.0)", "2");
+  assert_template_format("$(round 0.5)", "1");
+  assert_template_format("$(floor 0.7)", "0");
+  assert_template_format("$(ceil 0.2)", "1");
```
[optional]
Could you please extend the UT test set with scientific notation number parsing cases. Like `$(+ 1e-1 0)` (it already supports those).

smortex replied:
Sure, will use `5e-1` because, you know IEEE-754 cannot represent a lot of numbers with 20 digits precision: actual=0.10000000000000000555, expected=0.10000000000000000000 :smile:

Kokan commented:
I tried to summarise the result of that conversation:

A short summary of the discussion about this topic:
===

## Re/defining the arithmetic operations:

The way how operation should be defined: like the one you proposed in your PR, the one in the issue( new operations for float).
The conclusion on that was that your proposal is the solution we want to have. (Mainly because of the next point, and because it is natural on some level.)

## Impact on current configurations

With your version it is not an issue, as float was not supported before and operations with float resulted invalid/unwanted result (like *NaN*)

## Storage method/1

There is a recurring idea that nv-pairs should have more sophisticated way to store different types (other then the string representation).
This discussion was halted as it is kinda complex, and that would explode the work needed to be done now (also could be splitted as well).

## Storage method/2

Your proposal stores the float in a string representation. (this is somewhat connected to the next topic)
That introduces a precision syslog-ng could handle, which must be defined and compatible what *float*/*double* precision could do.


## Operations error propagation

A brief discussion about the errors each operation could introduce, or just the converting into string.
The conclusion on this was that syslog-ng is not a scientific tool, and as such the error propagation is a best effort basis.
Meaning if some issue raises on the matter, we can open a discussion to solve or state that won't solve such an issue;
as likely such discussion will never come up.

# What it means for this PR:

* The way you defined the operation extension is OK
* The precision/storage need some improvement

Kokan commented on modules/basicfuncs/numeric-funcs.c:
```diff
@@ -56,83 +111,204 @@ tf_num_parse(gint argc, GString *argv[],
+  else
+    {
+      gdouble res = value_as_double(n) - value_as_double(m);
+      g_string_append_printf(result, "%f", res);
```
Regarding the precision, while the `double` could hold a nice range. Saving that into that as string has impact on memory usage. It would be really nice to know the usecases it could come up.

I would suggest an arbitrary number to be 20 fractal digit.
This is based on assumption that *float* is used for time (like fraction of seconds), and the current limitation of such measurement according to wikipedia is somewhere near the `10e-17` range. [Atomic clock](https://en.wikipedia.org/wiki/Atomic_clock#Secondary_representations_of_the_second)
Note: this is a really big reduction specially for `double`, but keep in mind that each new digit is stored as string
I am happy with other values, this is more like a conversation starter.

What precision would you suggest ?

smortex replied:
My use case is related to milliseconds.  Because of the roundings, I assumed a 6 digits precision was okayish (as you say, syslog-ng is not a scientific tool), but for sure, more precision is probably better and 20 digits seems a reasonable choice.

smortex commented:
I added a bunch of commits to fix the reported issues (many thanks, the level of details is just perfect!).

In order to make reviewing easier, I have not touched previous commits, and will only squash them when you are okay with it.

kira-syslogng commented:
Build SUCCESS

smortex commented:
@Kokan, just to be sure, can you please confirm that your approval has value of "ok for squashing the commits and moving on to the next step"?

Kokan commented:
@smortex yes for the squash; but if by next step you mean merging the PR, then no. as that requires at least two approver.

smortex commented:
Sure! By "next step" I meant further work on the PR ;-)

smortex commented:
I reordered the commits, fixed and squashed the commits after rebasing on top of the master branch.  I pushed the two modified commits one by one in order to trigger two builds.  Let's see how is goes :wink:

kira-syslogng commented:
Build SUCCESS

smortex commented:
CI on Travis-CI fails on OSX because unit tests are disabled.  This seems to be caused by criterion not being installed in a previous stage of the build.  Transient failure?

Kokan commented:
Yes, after restarting that is usually fixed. Sometimes brew cannot install criterion :/

smortex commented:
Cool!  Thank you for re-triggering the build.

Do you think that I should add a way to ask `round` to accept an optional argument (defaulting to 0) for setting the precision in decimal digits:

```
(round 2.123456)   #=> "2"
(round 2.123456 3) #=> "2.123"
```

I am thinking that making it possible to have an integer or a float with 20 decimal digits is maybe a bit restrictive for some use cases (in my case, I manipulate data with 3 digits precision, so it does not make sense to have more than 3 digits in my final result)?

Kokan commented:
I agree on supporting the number of digit precision for `round`, and as well with the default to 0.
I would say it is fine to expand this PR until your usecase is completely covered; unless you want/could split this PR up into smaller chunks that make sense. In this case I would include the change regarding to `round`.

kira-syslogng commented:
Build FAILURE

smortex commented:
Thanks!  I added the optional precision argument to round, and reached a complete coverage of my use-case.

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

Kokan commented on modules/basicfuncs/tests/test_basicfuncs.c:
```diff
@@ -219,6 +219,23 @@ Test(basicfuncs, test_numeric_funcs)
+
+  assert_template_format("$(+ 5e-1 0)", "0.50000000000000000000");
+
+  assert_template_format("$(round 2.0)", "2");
```
[optional] I would extend the testcase with
```c
assert_template_format("$(round 2 -1)", "NaN");
assert_template_format("$(round 2 21)", "NaN");
assert_template_format("$(round 2 0)", "2");
assert_template_format("$(round 2 20)", "2.00000000000000000000");
```

smortex commented on modules/basicfuncs/numeric-funcs.c:
```diff
@@ -56,83 +137,248 @@ tf_num_parse(gint argc, GString *argv[],
+    {
+      if (!parse_dec_number(argv[1]->str, &precision))
+        {
+          msg_error("Parsing failed, template function's second argument is not a number",
```
Oops, this was completely unintentional, the rest of the code use `msg_debug()`, I guess I copy-pasted some code too far away and did not see that the function was not the same :confused:

smortex commented on modules/basicfuncs/numeric-funcs.c:
```diff
@@ -21,11 +21,92 @@
+{
+  char format[6];
+  snprintf(format, sizeof(format), "%%.%df", precision);
+  g_string_append_printf(result, format, number_as_double(n));
```
I realized this after committing my first batch of changes, and amended this old code.

smortex commented on modules/basicfuncs/numeric-funcs.c:
```diff
@@ -87,19 +106,24 @@ parse_integer_or_float(const char *str, Number *number)
+      /*
+       * This special case is required because:
+       * round(0.5)          #=> 1
+       * printf("%.0f", 0.5) #=> 0
```
I am still investigating the reason of this, and hope to simplify this code when I fully understand what is going on here.

smortex replied:
The [POSIX specification of the 'f' format string](https://pubs.opengroup.org/onlinepubs/009695399/functions/fprintf.html) says (emphasis by myself):

> f, F
>     The double argument shall be converted to decimal notation in the style "[-]ddd.ddd", where the number of digits after the radix character is equal to the precision specification. If the precision is missing, it shall be taken as 6; if the precision is explicitly zero and no '#' flag is present, no radix character shall appear. If a radix character appears, at least one digit appears before it. **The low-order digit shall be rounded in an implementation-defined manner.**

So le'ts use `round(3)` when possible, and rely on the implementation-defined rounding otherwise.

smortex replied:
Do you have an opinion about this?

1) Is this okay as it is currently?
2) Should we always round using %f (that is the test suite needs to be updated because 0.5 is rounded to 0)?
3) Should we always use round(3) (mutiply the number by 10^precision, do the rounding, divide by 10^precision and display this result with %f and precision digits)?

More and more I think 3 is the best way to go because of the consistency of the roundings, but 1 is ways more simple and maybe acceptable for a logging software. 2 is hybrid and inconsistent, so maybe it's the worst possibility.

What do you think?

Kokan replied:
I can agree on the 3rd option, as precision cannot be a conserve while it is limited to 20 digits from the 300 something `gdouble` has, in addition to the reasons you wrote.

Kokan replied:
I left a note for the wip patch you linked.

smortex commented:
I added a bunch of commit that address the problems you spotted.  Thanks!

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

Kokan commented:
@smortex please feel free to fixup the changes in the next step (if you agree with my comment on your commit), and from my side it is good to go.

kira-syslogng commented:
Build SUCCESS

smortex commented:
Thanks @Kokan I squashed the last commits into a single one.  It looks like the "default" kira build fails, but I cannot access the system.  Is there something I can fix myself?

Kokan commented:
@smortex please ignore the *default* kira.

bazsi commented:
That job is work in progress, so ignore it please. I have disabled it for
now.

Kókai Péter <notifications@github.com> ezt írta (időpont: 2019. jún. 6., Cs
23:05):

> @smortex <https://github.com/smortex> please ignore the *default* kira.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/2742?email_source=notifications&email_token=AAFOK5V6CK5QPJQDPDWWY5LPZF3ZTA5CNFSM4HOO7YU2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXEFDEQ#issuecomment-499667346>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFOK5VOZF5DRLKEHAASIP3PZF3ZTANCNFSM4HOO7YUQ>
> .
>

Kokan commented:
@kira-syslogng reteset this please
(@smortex I'll do some stuff here to make default green)

Kokan commented:
@kira-syslogng retest this please

kira-syslogng commented:
Build FAILURE

Kokan commented:
@kira-syslogng retest this please

kira-syslogng commented:
Build SUCCESS
